gemini/responsible-ai/react_rag_attacks_mitigations_examples.ipynb

```diff
@@ -0,0 +1,1086 @@
+    "            generation_config=generation_config,\n",
+    "        )\n",
+    "\n",
+    "        response_last_lines = \"\\n\".join(response.text.splitlines()[-3:])\n",
+    "\n",
+    "        if \"WAITING\" in response_last_lines:\n",
+    "            action, action_input = extract_action(response_last_lines)\n",
+    "\n",
+    "            if action == \"weather_city\":\n",
+    "                observation = weather_city(action_input)\n",
+    "            elif action == \"order_store\":\n",
+    "                observation = order_store(action_input)\n",
+    "\n",
+    "            agent_scratchpad = (\n",
+    "                agent_scratchpad + response.text + f\"Observation: {observation}\\n\"\n",
+    "            )\n",
+    "        else:\n",
+    "            agent_scratchpad = agent_scratchpad + response.text\n",
+    "            break\n",
+    "\n",
+    "    return agent_scratchpad"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "y4ogNzD1tYVT",
+   "metadata": {
+    "id": "y4ogNzD1tYVT"
+   },
+   "outputs": [],
+   "source": [
+    "def order_store(item):  # type: ignore[no-redef]\n",
+    "    \"\"\"Concludes with a fictive order at online store\"\"\"\n",
+    "\n",
+    "    print(\n",
+    "        f\">>> Action: order_store, Input: {item}, Return:Placed in Basket. Please login and confirm the order\"\n",
+    "    )\n",
+    "    return f\"Order placed in Basket. Please login and confirm {item}\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "elomuRURtkO5",
+   "metadata": {
+    "id": "elomuRURtkO5"
+   },
+   "outputs": [],
+   "source": [
+    "print(chat(\"What is the weather in Munich ?\"))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "PeHOnFxrbfo8",
+   "metadata": {
+    "id": "PeHOnFxrbfo8"
+   },
+   "source": [
+    "## Retrieval-augmented generation (RAG)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "j7ZkutnfYTlO",
+   "metadata": {
+    "id": "j7ZkutnfYTlO"
+   },
+   "source": [
+    "![rag.png](https://storage.googleapis.com/github-repo/responsible-ai/intro_genai_security/rag.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "oQ0OzLtZizCu",
+   "metadata": {
+    "id": "oQ0OzLtZizCu"
+   },
+   "source": [
+    "*Let us assume the company has a lot of historically generated PDF files from different tools. The company wants to use RAG to get more insight and customer value out of the documents.*\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "kp3YQe3w6rQj",
+   "metadata": {
+    "id": "kp3YQe3w6rQj"
+   },
+   "source": [
+    "**Upload** to your Colab default folder the PDF test files\n",
+    "- [Beyond41.pdf](https://storage.googleapis.com/github-repo/responsible-ai/intro_genai_security/Beyond41.pdf)\n",
+    "- [Beyond41mal.pdf](https://storage.googleapis.com/github-repo/responsible-ai/intro_genai_security/Beyond41mal.pdf)"
```

Comment:
Can you rewrite this to have the user download using `gsutil`? Not all users will be using Colab.

Reply:
fixed